File: architecture/data/src/main/java/com/pacific/data/HttpUtil.java
package com.pacific.data;

import android.text.TextUtils;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Map;

import okhttp3.Call;
import okhttp3.OkHttpClient;
import okio.ByteString;

public class HttpUtil {

    private HttpUtil() {
        throw new UnsupportedOperationException();
    }

    public static String md5(String source) {
        if (TextUtils.isEmpty(source)) {
            return source;
        }
        return ByteString.encodeUtf8(source).md5().hex();
    }

    @SuppressWarnings("unchecked")
    public static Map<String, String> toMap(Object obj, Moshi moshi) {
        if (moshi == null) {
            moshi = new Moshi.Builder().build();
        }
        JsonAdapter<Object> adapter = moshi.adapter(Object.class);
        Object jsonStructure = adapter.toJsonValue(obj);
        return (Map<String, String>) jsonStructure;
    }

    public static String toJson(Object obj, Moshi moshi, Type type) {
        if (moshi == null) {
            moshi = new Moshi.Builder().build();
        }
        return moshi.adapter(type).lenient().toJson(obj);
    }

    @SuppressWarnings("unchecked")
    public static <T> T fromJson(String json, Moshi moshi, Type type) throws IOException {
        if (moshi == null) {
            moshi = new Moshi.Builder().build();
        }
        return (T) moshi.adapter(type).lenient().fromJson(json);
    }

    public static byte[] toJsonByteArray(Object obj, Moshi moshi, Type type) {
        return string2ByteArray(toJson(obj, moshi, type));
    }

    @SuppressWarnings("unchecked")
    public static <T> T fromJsonByteArray(byte[] jsonByteArray, Moshi moshi, Type type)
            throws IOException {
        return fromJson(byteArray2String(jsonByteArray), moshi, type);
    }

    public static byte[] string2ByteArray(String value) {
        return ByteString.encodeUtf8(value).toByteArray();
    }

    public static String byteArray2String(byte[] bytes) {
        return ByteString.of(bytes, 0, bytes.length).utf8();
    }

    public void cancelRequest(OkHttpClient client, Object tag) {
        for (Call call : client.dispatcher().queuedCalls()) {
            if (tag.equals(call.request().tag())) {
                call.cancel();
            }
        }
        for (Call call : client.dispatcher().runningCalls()) {
            if (tag.equals(call.request().tag())) {
                call.cancel();
            }
        }
    }
}

File: architecture/arch-kotlin/src/main/java/com/pacific/arch/kotlin/data/HttpUtil.kt
package com.pacific.arch.kotlin.data

import com.squareup.moshi.KotlinJsonAdapterFactory
import com.squareup.moshi.Moshi
import okhttp3.OkHttpClient
import okio.ByteString
import java.io.IOException
import java.lang.reflect.Type

fun cancelOkhttp3Request(okHttpClient: OkHttpClient, tag: Any) {
    okHttpClient.dispatcher().queuedCalls()
            .filter { tag == it.request().tag() }
            .forEach { it.cancel() }
    okHttpClient.dispatcher().runningCalls()
            .filter { tag == it.request().tag() }
            .forEach { it.cancel() }
}

fun md5(src: String) = ByteString.encodeUtf8(src).md5().hex()!!

@JvmOverloads
@Suppress("UNCHECKED_CAST")
fun toMap(obj: Any, moshi: Moshi? = null): Map<String, String> {
    val mMoshi = moshi ?: Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
    val adapter = mMoshi!!.adapter(Any::class.java)
    return adapter.toJsonValue(obj) as Map<String, String>
}

@JvmOverloads
fun toJson(obj: Any, type: Type, moshi: Moshi? = null): String {
    val mMoshi = moshi ?: Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
    return mMoshi.adapter<Any>(type).lenient().toJson(obj)
}

@JvmOverloads
@Throws(IOException::class)
@Suppress("UNCHECKED_CAST")
fun <T> fromJson(json: String, type: Type, moshi: Moshi? = null): T {
    val mMoshi = moshi ?: Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
    return mMoshi!!.adapter<T>(type).lenient().fromJson(json)!!
}

@JvmOverloads
fun toByteArrayJson(obj: Any, type: Type, moshi: Moshi? = null): ByteArray {
    return string2ByteArray(toJson(obj, type, moshi))
}

@JvmOverloads
@Throws(IOException::class)
fun <T> fromByteArrayJson(byteArray: ByteArray, type: Type, moshi: Moshi? = null): T {
    return fromJson(byteArray2String(byteArray), type, moshi)
}

fun string2ByteArray(src: String): ByteArray {
    return ByteString.encodeUtf8(src).toByteArray()
}

fun byteArray2String(bytes: ByteArray): String {
    return ByteString.of(bytes, 0, bytes.size).utf8()
}